pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/automaton/Automaton.java

```diff
@@ -0,0 +1,652 @@
+    boolean b = _allowMutation;
+    _allowMutation = flag;
+    return b;
+  }
+
+  /**
+   * Assigns consecutive numbers to the given states.
+   */
+  static void setStateNumbers(Set<State> states) {
+    if (states.size() == Integer.MAX_VALUE) {
+      throw new IllegalArgumentException("number of states exceeded Integer.MAX_VALUE");
+    }
+    int number = 0;
+    for (State s : states) {
+      s._number = number++;
+    }
+  }
+
+  /**
+   * Returns a sorted array of transitions for each state (and sets state numbers).
+   */
+  static Transition[][] getSortedTransitions(Set<State> states) {
+    setStateNumbers(states);
+    Transition[][] transitions = new Transition[states.size()][];
+    for (State s : states) {
+      transitions[s._number] = s.getSortedTransitionArray(false);
+    }
+    return transitions;
+  }
+
+  /**
+   * See {@link MinimizationOperations#minimize(Automaton)}.
+   * Returns the automaton being given as argument.
+   */
+  public static Automaton minimize(Automaton a) {
+    a.minimize();
+    return a;
+  }
+
+  void checkMinimizeAlways() {
+    if (_minimizeAlways) {
+      minimize();
+    }
+  }
+
+  boolean isSingleton() {
+    return _singleton != null;
+  }
+
+  /**
+   * Gets initial state.
+   * @return state
+   */
+  public State getInitialState() {
+    expandSingleton();
+    return _initial;
+  }
+
+  /**
+   * Sets initial state.
+   * @param s state
+   */
+  public void setInitialState(State s) {
+    _initial = s;
+    _singleton = null;
+  }
+
+  /**
+   * Returns the set of states that are reachable from the initial state.
+   * @return set of {@link State} objects
+   */
+  public Set<State> getStates() {
+    expandSingleton();
+    Set<State> visited;
+
+    visited = new HashSet<>();
+    LinkedList<State> worklist = new LinkedList<State>();
+    worklist.add(_initial);
+    visited.add(_initial);
+    while (worklist.size() > 0) {
+      State s = worklist.removeFirst();
+      Collection<Transition> tr;
+
+      tr = s._transitionSet;
+      for (Transition t : tr) {
+        if (!visited.contains(t._to)) {
+          visited.add(t._to);
+          worklist.add(t._to);
+        }
+      }
+    }
+    return visited;
+  }
+
+  /**
+   * Returns the set of reachable accept states.
+   * @return set of {@link State} objects
+   */
+  public Set<State> getAcceptStates() {
+    expandSingleton();
+    HashSet<State> accepts = new HashSet<State>();
+    HashSet<State> visited = new HashSet<State>();
+    LinkedList<State> worklist = new LinkedList<State>();
+    worklist.add(_initial);
+    visited.add(_initial);
+    while (worklist.size() > 0) {
+      State s = worklist.removeFirst();
+      if (s._accept) {
+        accepts.add(s);
+      }
+      for (Transition t : s._transitionSet) {
+        if (!visited.contains(t._to)) {
+          visited.add(t._to);
+          worklist.add(t._to);
+        }
+      }
+    }
+    return accepts;
```

Comment:
Have you considered assigning each state an incremental identifier unique within the scope of an automaton, then you can do a lot of this logic with bitsets?

Reply:
In fact you have this already: `_number`